rbmc-cfam-daemon/src/bmc_cfam.hpp

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include "cfam_fields.hpp"
+
+#include <algorithm>
+#include <map>
+#include <set>
+
+/**
+ * @class BMCCFAM
+ *
+ * The base class representing a CFAM-S on a BMC card, where the four
+ * scratchpad registers in the Gemini Mailbox block of the CFAM are
+ * used to store redundant BMC information.
+ *
+ * Children of this base class are local CFAM and sibling CFAM classes.
+ */
+class BMCCFAM
+{
+  public:
+    /**
+     * @brief The fields stored in the mailbox scratchpad registers
+     * in CFAM-S chips off of a BMC.
+     */
+    enum class Field
+    {
+        apiVersion,
+        bmcPosition,
+        role,
+        redundancyEnabled,
+        failoversPaused,
+        provisioned,
+        bmcState,
+        siblingCommsOK,
+        heartbeat,
+        fwVersion
+    };
+
+    /**
+     * @brief The register, starting bit, and length of the fields.
+     */
+    inline static const std::map<Field, cfam::FieldMetadata> cfamFields{
+        {Field::apiVersion, {cfam::ScratchPadReg::one, 0, 8}},
+        {Field::bmcPosition, {cfam::ScratchPadReg::one, 8, 1}},
+        {Field::role, {cfam::ScratchPadReg::one, 9, 2}},
+        {Field::redundancyEnabled, {cfam::ScratchPadReg::one, 11, 1}},
+        {Field::failoversPaused, {cfam::ScratchPadReg::one, 12, 1}},
+        {Field::provisioned, {cfam::ScratchPadReg::one, 13, 1}},
+        {Field::bmcState, {cfam::ScratchPadReg::one, 14, 3}},
+        {Field::siblingCommsOK, {cfam::ScratchPadReg::one, 17, 1}},
+        {Field::heartbeat, {cfam::ScratchPadReg::one, 24, 8}},
+        {Field::fwVersion, {cfam::ScratchPadReg::two, 0, 32}}};
+
+    BMCCFAM() = default;
+    virtual ~BMCCFAM() = default;
+    BMCCFAM(const BMCCFAM&) = delete;
+    BMCCFAM& operator=(const BMCCFAM&) = delete;
+    BMCCFAM(BMCCFAM&&) = delete;
+    BMCCFAM& operator=(BMCCFAM&&) = delete;
+
+    /**
+     * @brief Of the four possible scratchpad registers, returns the ones
+     *        that are used for fields.
+     *
+     *  @return The set of registers
+     */
+    inline static std::set<cfam::ScratchPadReg> getUsedRegs()
```

Comment:
Yea, sorry I have other commits ready that use this, but didn't want to flood you with too much at once.

it will eventually be used by a derived class that calls it in the ctor:

```
    SiblingCFAM(size_t link, SysFS& sysfs) :
        BMCCFAM(link, sysfs), usedRegs(BMCCFAM::getUsedRegs())
    {}
```